Mathlib/Tactic/TFAE.lean

```diff
@@ -159,62 +219,72 @@ def proveTFAE (is : List ℕ) (l : Q(List Prop)) : MetaM Q(TFAE $l) := do
+    | _ => Macro.throwUnsupported
+    return .mkSimple <| String.intercalate "_" ["tfae", s!"{i.getNat}", arr, s!"{j.getNat}"]
+  | _ => Macro.throwUnsupported
 
 /-- Turn syntax for a given index into a natural number, as long as it lies between `1` and
 `maxIndex`. -/
-def elabIndex (i : TSyntax `num) (maxIndex : ℕ) : TacticM ℕ := do
+def elabIndex (i : TSyntax `num) (maxIndex : ℕ) : MetaM ℕ := do
   let i' := i.getNat
-  unless Nat.ble 1 i' && Nat.ble i' maxIndex do
-    throwError "{i} must be between 1 and {maxIndex}"
+  unless 1 ≤ i' && i' ≤ maxIndex do
+    throwErrorAt i "{i} must be between 1 and {maxIndex}"
   return i'
 
-/-- Construct an expression for the type `Pj → Pi`, `Pi → Pj`, or `Pi ↔ Pj` given expressions
-`Pi Pj : Q(Prop)` and `impArrow` syntax `arr`, depending on whether `arr` is `←`, `→`, or `↔`
-respectively. -/
-def mkImplType (Pi : Q(Prop)) (arr : TSyntax ``impArrow) (Pj : Q(Prop)) : MetaM Q(Prop) := do
-  match arr with
-  | `(impArrow| ← ) => pure q($Pj → $Pi)
-  | `(impArrow| → ) => pure q($Pi → $Pj)
-  | `(impArrow| ↔ ) => pure q($Pi ↔ $Pj)
-  | _ => throwErrorAt arr "expected '←', '→', or '↔'"
-
 /-! # Tactic implementation -/
 
-elab_rules : tactic
-| `(tactic| tfae_have $[$h:ident : ]? $i:num $arr:impArrow $j:num) => do
-  let goal ← getMainGoal
-  goal.withContext do
-    let (_, tfaeList) ← getTFAEList (← goal.getType)
-    let l₀ := tfaeList.length
-    let i' ← elabIndex i l₀
-    let j' ← elabIndex j l₀
+/-- Accesses the propositions at indices `i` and `j` of `tfaeList`, and constructs the expression
+`Pi <arr> Pj`, which will be the type of our `tfae_have` hypothesis -/
+def elabTFAEType (tfaeList : List Q(Prop)) : TSyntax ``tfaeType → TermElabM Expr
+  | stx@`(tfaeType|$i:num $arr:impArrow $j:num) => do
+    let l := tfaeList.length
+    let i' ← elabIndex i l
+    let j' ← elabIndex j l
     let Pi := tfaeList.get! (i'-1)
     let Pj := tfaeList.get! (j'-1)
-    let type ← mkImplType Pi arr Pj
-    let (goal1, goal2) ← tfaeHaveCore goal h i j arr type
-    replaceMainGoal [goal1, goal2]
+    Term.addTermInfo' i Pi q(Prop)
+    Term.addTermInfo' j Pj q(Prop)
+    match arr with
+    | `(impArrow| ← ) => Term.addTermInfo stx q($Pj → $Pi) q(Prop)
+    | `(impArrow| → ) => Term.addTermInfo stx q($Pi → $Pj) q(Prop)
+    | `(impArrow| ↔ ) => Term.addTermInfo stx q($Pi ↔ $Pj) q(Prop)
+    | _ => throwUnsupportedSyntax
+  | _ => throwUnsupportedSyntax
+
+/- Convert `tfae_have i <arr> j ...` to `tfae_have tfae_i_arr_j : i <arr> j ...`. See
+`expandHave`, which is responsible for inserting `this` in `have : A := ...`. -/
+macro_rules
+| `(tfaeHave|tfae_have $hy:hygieneInfo $t:tfaeType := $val) => do
+  let id := HygieneInfo.mkIdent hy (← mkTFAEId t) (canonical := true)
+  `(tfaeHave|tfae_have $id : $t := $val)
+| `(tfaeHave|tfae_have $hy:hygieneInfo $t:tfaeType $alts:matchAlts) => do
+  let id := HygieneInfo.mkIdent hy (← mkTFAEId t) (canonical := true)
+  `(tfaeHave|tfae_have $id : $t $alts)
```

Comment:
I think it would be better to do this inside the elab and use `Lean.Meta.mkFreshBinderNameForTactic` in order to make the tactic hygienic.

I.e. the binder name should probably only be accessible if `set_option tactic.hygienic false`

Reply:
Hmm, I'm not sure—`tfae_have` is specifically meant to mimic `have` when the binder name is not provided, which means producing an accessible name like `tfae_2_iff_3` in analogy to producing the accessible name `this`.

(This is also how it currently behaves, so I'd be wary of breaking it!)

Reply:
ok that's a fair point!